rfcs/0006-merge-operator.md

```diff
@@ -0,0 +1,450 @@
+
+Aside from this, `Merge` rows can follow the same structure as `Value` rows.
+
+This change is backwards compatible.
+
+### Block Format
+
+The block format remains unchanged. It's useful to hightlight that merge operands are also stored in the bloom filters, so they have the same filtering guarantees as regular values.
+
+### Read Path
+
+Merges are expected to be read using the standard `get` API.
+
+For each key, SlateDB maintains a history of operations, which is subject to compaction.
+
+Let's denote each operation as `OPi`. A key (`K`) that has experienced `n` changes looks like this logically (physically, the changes could be spread across the memtable and multiple SST files):
+
+```
+K:    OP1    OP2    OP3    ...    OPn
+
+earliest --------------------> latest
+```
+
+Currently, during a read operation, SlateDB always observes the latest known state of the database. In other words, it only needs to look up the latest operation. If it's a `Value`, it simply returns it, and if it's a `Tombstone`, it returns `None`. All previous operations can be ignored and will be removed by compaction at some point.
+
+```
+K:    OP1    OP2    OP3    ...    OPn
+                                   ^
+                                   |
+                                  GET
+```
+
+With the addition of a **Merge Operand**, it may be necessary to look backwards at previous values, up to the point where we encounter either a `Value` or a `Tombstone`. Everything beyond that point can be ignored.
+
+```
+K:    OP1    OP2    OP3    OP4    OPn
+            Value  Merge  Merge  Merge
+                                   ^
+                                   |
+                                  GET
+```
+
+
+In the above example, `get` should essentially return something like:
+
+```
+Merge(OP2, Merge(OP3, Merge(OP4, OPn)))
+```
+
+### Write Path & Compactions
+
+During compaction, we continuously reduce the history by merging entries according to the rules described above. This process performs partial merges along the way, combining merge operands with their base values when possible. We also merge entries in the memtable as they arrive to minimize unnecessary history storage.
+
+Compaction operates on a set of source SST files and writes to a destination SST file. The process reads all source SSTs sequentially and produces a single destination SST containing the merged results.
+
+The table below shows how different combinations of previous and new values are handled during compaction:
+
+| Previous  | New       | Result    |
+|-----------|-----------|-----------|
+| tombstone | value     | value     |
+| tombstone | merge     | value     |
+| tombstone | tombstone | tombstone |
+| value     | merge     | value     |
+| value     | value     | value     |
+| value     | tombstone | tombstone |
+| merge     | merge     | merge     |
+| merge     | value     | value     |
+| merge     | tombstone | tombstone |
+
+#### Snapshot Isolation
```

Comment:
Is the below scenario valid?
Transaction1:
Merge1
Merge2
Transaction2:
Merge3
Commit
Transaction1:
Merge 4.

1. Until Transaction1 commits, I assume we wouldn't consider Merge1 and Merge2.
2. Can Transaction1 commit at all? Would it fail?

Reply:
That scenario is perfectly fine unless you do a read operation in between (this has same semantics as `PUT` operation). The only caveat is that you can't eagerly merge everything as described below.